From: SeongJae Park <sj@kernel.org>
Subject: mm/damon/core: implement damon_initialized() function
Date: Mon, 15 Sep 2025 20:35:05 -0700

Patch series "mm/damon: define and use DAMON initialization check
function".

If DAMON is used when it is not yet successfully initialized, the caller
could crash.  The DAMON core layer does not providing a reliable way to
see if it is successfully initialized and therefore ready to be used.  As
a result, DAMON API callers are implementing their own hacks to see this. 
The hacks simply assume DAMON should be ready on module init time.  It is
not reliable as DAMON initialization can indeed fail if KMEM_CACHE()
fails, and difficult to maintain as those are duplicates.  Implement a
core layer API function for better reliability and maintainability to
replace the hacks with followup commits.


This patch (of 7):

If DAMON is tried to be used when it is not yet successfully initialized,
the caller could be crashed.  DAMON core layer is not providing a reliable
way to see if it is successfully initialized and therefore ready to be
used, though.  As a result, DAMON API callers are implementing their own
hacks to see it.  The hacks simply assume DAMON should be ready on module
init time.  It is not reliable as DAMON initialization can indeed fail if
KMEM_CACHE() fails, and difficult to maintain as those are duplicates. 
Implement a core layer API function for better reliability and
maintainability to replace the hacks with followup commits.

Link: https://lkml.kernel.org/r/20250916033511.116366-2-sj@kernel.org
Link: https://lkml.kernel.org/r/20250916033511.116366-2-sj@kernel.org
Signed-off-by: SeongJae Park <sj@kernel.org>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 include/linux/damon.h |    1 +
 mm/damon/core.c       |   10 ++++++++++
 2 files changed, 11 insertions(+)

--- a/include/linux/damon.h~mm-damon-core-implement-damon_initialized-function
+++ a/include/linux/damon.h
@@ -938,6 +938,7 @@ static inline unsigned int damon_max_nr_
 }
 
 
+bool damon_initialized(void);
 int damon_start(struct damon_ctx **ctxs, int nr_ctxs, bool exclusive);
 int damon_stop(struct damon_ctx **ctxs, int nr_ctxs);
 bool damon_is_running(struct damon_ctx *ctx);
--- a/mm/damon/core.c~mm-damon-core-implement-damon_initialized-function
+++ a/mm/damon/core.c
@@ -2863,6 +2863,16 @@ void damon_update_region_access_rate(str
 		r->nr_accesses++;
 }
 
+/**
+ * damon_initialized() - Return if DAMON is ready to be used.
+ *
+ * Return: true if DAMON is ready to be used, false otherwise.
+ */
+bool damon_initialized(void)
+{
+	return damon_region_cache != NULL;
+}
+
 static int __init damon_init(void)
 {
 	damon_region_cache = KMEM_CACHE(damon_region, 0);
_
